crypto/x509/v3_ncons.c

```diff
@@ -796,6 +796,14 @@ static int nc_uri(ASN1_IA5STRING *uri, ASN1_IA5STRING *base)
 
     /* Determine length of hostname part of URI */
 
+    /*
+     * First look if there is a userinfo; if there is, skip it
+     * to get to the actual start of the hostname.
+     */
+    p = ia5memchr(uri, hostptr, '@');
+    if (p != NULL)
+        hostptr = p + 1;
```

Comment:
The parsing of the URI scheme + authority above sets a bad precedent.  The code should really have ensured that the URI starts with a valid *scheme* which is then followed by `://` which does not first appear somewhere else in the URI and then below work with a copy of just the authority part before dropping any *userinfo* part and subsequent `@`.

The good news is that OpenSSL now provides `OSSL_parse_url()` which should take care of both, by parsing the URL into all its constituent parts, you want just the *scheme* (making sure it is non-empty) and *host* parts.

The bad news is that looking `OSSL_parse_url()` I see some defects, its *scheme* parser does not enforce the requisite restrictions on the *scheme* character set, and be a bit sloppy in other ways.

Nevertheless, use that function, and a separate issue/PR can be opened to improve its implementation.

Reply:
Just to reassure you that this has not been abandonded.... I'm still waiting on the okay from our legal team and until I get it, I'm afraid I can't work on this.